src/FwdState.cc

```diff
@@ -565,6 +566,9 @@ FwdState::complete()
 
     logReplyStatus(n_tries, replyStatus);
 
+    // will already be false if complete() was called before/without dispatch()
+    waitingForDispatched = false;
```

Comment:
I noticed that we reset `waitingForDispatched` before resetting `serverConn` and `destinationReceipt` in two places. Probably we could do likewise in `complete()`, moving the 570 line under the first 'if' because `waitingForDispatched` should not matter in the 'else' terminating with stopAndDestroy(). If so, should not we create a single method for this cleanup?
Also, since dispatch() is always called before syncWithServerConn() (where `serverConn` is assigned), should not we also move there `waitingForDispatched` setting? In other words, if the point since then the destination is considered to be 'used' is after FwdState becomes the owner of `serverConn`, we should probably assign the flag there (and adjust its name, if needed).

Reply:
Thank you for a useful review. I started responding yesterday, but could not finish at that time. I should have removed the M-cleared-for-merge label! We will do a followup PR to reflect your ideas.

Reply:
> Eduard: I noticed that we reset `waitingForDispatched` before resetting `serverConn` and `destinationReceipt` in two places. Probably we could do likewise in `complete()`, moving the 570 line under the first 'if' because `waitingForDispatched` should not matter in the 'else' terminating with stopAndDestroy(). If so, should not we create a single method for this cleanup?

I think the answer is "yes, with minor modifications[^1]" for the current code, but I think we should delay this cleanup. If we create a single method for the connection/waiting cleanup now, we will go a step back towards tying connection ownership (serverConn) and waitingForDispatched state. That misleading association is what led to the bug this PR fixed.

[^1]: If we adjust complete() along these lines, we should clear both waitingForDispatched and serverConn unconditionally, not inside the first `if` branch. We stop waiting for the dispatched transaction in complete() unconditionally _and_ should get rid of any connection leftovers unconditionally.

> Eduard: Also, since dispatch() is always called before syncWithServerConn() (where `serverConn` is assigned), should not we also move there `waitingForDispatched` setting? In other words, if the point since then the destination is considered to be 'used' is after FwdState becomes the owner of `serverConn`, we should probably assign the flag there (and adjust its name, if needed).

Technically, you are right -- the code always syncs the server connection right before dispatch() -- but that syncing alone does not really place FwdState in a true waiting state. In fact, we currently set waitingForDispatched too _early_ -- our dispatch() method may not actually enter the waiting state either (e.g., in `request->flags.sslPeek` case and in the `default` getScheme() case)! The code "works", AFAICT, but it is misleading/messy. Another problem to fix.

> Alex: We will do a followup PR to reflect your ideas.

I think all this cleanup should happen when or after we change FwdState to stop owning the to-server connection on the way to dispatch(). The connection should probably become the dispatch() parameter, but more (and more complex!) changes are required to fully get rid of the serverConn member and associated problems.

Thank you again for your review -- it exposed more/new problems in the current code and will help us clean it up better. It may take some time for us to clean this up properly, more than I originally expected, but we are on the right path AFAICT.